### ISSUE TITLE
[Usage Question] How to improve (reduce) the evaluation metrics when training complex system (Zeolites) MD data with forces?

### ISSUE THREAD

**Issue by Chenghan-Sun** [atomistic-machine-learning/schnetpack#416]
Hi SchNetPack development team,

I've been using SchNetPack as an alternative approach to the traditional descriptor-based molecular representations for a while, and I also did some research on the original published papers of SchNet as well as some recent applications of SchNet in various molecular prediction tasks. 

Recently, I'm trying to use SchNet representation to train on about 10K (specifically, 10344 images) MD data with both energies and forces on Chabazite Zeolite system (108 atoms per image, with 72O and 36Si). In this dataset, PBC is guaranteed, so even if the number of atoms per image is relatively large, that should be the only resource of complexity regarding to the relatively simple local environments. However, with a very finely tuned hyper-parameters searching procedure, I'm still very struggled with training with forces (i.e., to achieve a low MAEs for decomposed forces in x-y-z directions). 

The finalized parameters and partitions I used is listed as:
```
Data Settings:
 num_train: 100 / 1000 / 10000 
 num_val: 20 / 100 / 344 
 num_test: the remaining dataset
 batch_size: 8

Model Settings:
 n_features: 128
 n_atom_basis: 128
 n_filters: 128
 n_gaussians: 60
 n_interactions: 6
 cutoff: 6.0
 force option: yes 

Training Settings:
 Device: cuda
 rho_tradeoff: 0.01
 learning rate: 0.001
 Patience: 5
 n_epochs: 300 
```

Here is some of my SchNetPack training results: 
- Task 1: randomly select 100 images as training set and 20 images as validation set:
```
MAE train = 0.365 eV; MAE validation = 0.356 eV (energy)
MAE train = 0.333 eV/A; MAE validation = 0.324  eV/A (forces at x axis )
MAE train = 0.340 eV/A; MAE validation = 0.306 eV/A (forces at y axis)
MAE train = 0.363 eV/A; MAE validation = 0.339 eV/A (forces at z axis)
```
- Task 2: randomly select 1000 images as training set and 100 images as validation set:
```
MAE train = 0.186 eV; MAE validation = 0.239 eV (energy)
MAE train = 0.283 eV/A; MAE validation = 0.298  eV/A (forces at x axis )
MAE train = 0.290 eV/A; MAE validation = 0.309 eV/A (forces at y axis)
MAE train = 0.318 eV/A; MAE validation = 0.334 eV/A (forces at z axis)
```
- Task 3: randomly select 10000 images as training set and the remaining 344 images as validation set:
```
MAE train = 0.067 eV; MAE validation = 0.112 eV (energy)
MAE train = 0.212 eV/A; MAE validation = 0.248  eV/A (forces at x axis )
MAE train = 0.218 eV/A; MAE validation = 0.248 eV/A (forces at y axis)
MAE train = 0.241 eV/A; MAE validation = 0.275 eV/A (forces at z axis)
```
All the training and validation loss functions plotted in scenarios above looks fairly smooth. So in summary, as I increase the size of training set, I do see improvements of MAEs for both energies and forces. The MAE of energies can be improved to between 0.1 eV and 0..15 eV, which is very acceptable considering the typical error of ab-Initio method (e.g. DFT) with +- 0.15 eV error. However, regarding the forces, the MAEs are generally too large. 

With the same setting of hyper-parameters, I finally did a full train-validation-test splits on my CHA database, following with 80-10-10 percent rule. Here is some more specific parity plots I obtained after reloaded the model on the test set: 
- Task 4: randomly select 8344 images as training set 1000 images as validation set, and the remaining 1000 images as test set:
```
MAE train = 0.079 eV; MAE validation = 0.119 eV; MAE test = 0.132 eV (energy)
MAE train = 0.218 eV/A; MAE validation = 0.256 eV/A; MAE test = 0.256 eV/A (forces at x axis )
MAE train = 0.224 eV/A; MAE validation = 0.262 eV/A; MAE test = 0.262 eV/A (forces at y axis)
MAE train = 0.246 eV/A; MAE validation = 0.285 eV/A; MAE test = 0.284 eV/A (forces at z axis)
```
Parity plots:
![energy_parity_plots_sns_ver](https://user-images.githubusercontent.com/43866006/171902522-0acc973f-e318-43f4-a2cb-3e5be8c651f4.png)
![force_x_parity_plots_sns_ver_mixedCHA_8344train_batch4](https://user-images.githubusercontent.com/43866006/171925637-60fc8174-3b32-4457-9444-7877c2f4620c.png)
![force_y_parity_plots_sns_ver_mixedCHA_8344train_batch4](https://user-images.githubusercontent.com/43866006/171925658-fe62dba8-601c-424f-9782-34fc79483074.png)
![force_z_parity_plots_sns_ver_mixedCHA_8344train_batch4](https://user-images.githubusercontent.com/43866006/171925676-6f1cb0a7-1908-412b-b82a-eeef9cf6cc90.png)

Loss functions: 
![loss_mixedCHA_model_v4_8344train](https://user-images.githubusercontent.com/43866006/171902999-92480e22-d496-4060-a20e-2bf9112c5204.png)

That's all my current progress now. One of my colleague used another ML package called DeepMDkit to train the same dataset, and been able to reduct the force MAEs to the 10^-2 scale. As I have been played with SchNetPack for a while, I hope to get some suggestions from the developer team regarding to making the forces training better. Any idea or suggestions will be appreciate! 

Best regards,
Chenghan Sun

**Comment by JordD04** [comment]
Hey Chenghan Sun,
I think you're experiencing similar problems to me (https://github.com/atomistic-machine-learning/schnetpack/issues/407). It may be worth visiting my thread to see if any of the recommendations made there are applicable to you.
I haven't had much luck but I haven't tried Stefaan's input file yet (doing that now).

**Comment by Stefaanhess** [comment]
Hey @Chenghan-Sun ,
I am unsure why you are getting a bad performance. Do you use the current master or dev branch? I just made a small tutorial for using the dev branch on a benchmark dataset with pbc and forces. You can find it as a comment in #407 . Could you try to follow the tutorial for the example data and run the training with the same config file on your own data by changing the data path? Afterwards you could upload the evaluation here and we can see if we can figure out what is going on.

**Comment by Stefaanhess** [comment]
Hey @Chenghan-Sun ,
did you test the example script from #407? If you are still using the current master, please switch to dev and try the example config. The config also uses the PaiNN representation, which seems to perform better than SchNet on pbc data. So this should improve your results.
I am closing this issue for now. Feel free to reopen if you have any more questions on this topic, or leave a comment at #407.

**Comment by Chenghan-Sun** [comment]
Hi @Stefaanhess and @JordD04 ,

Thank you very much for your kind advice and sorry for the late reply! 

So given @Stefaanhess 's previous concern about the version I used for training Zeolite images, I'll been busy with figuring out some PBC issue on the stable ver (the master branch). So I haven't really used the dev version actually. I thought the correct link posted in the dev version API: 
[https://schnetpack.readthedocs.io/en/dev/getstarted/firststeps.html](https://schnetpack.readthedocs.io/en/dev/getstarted/firststeps.html)
should be replaced by:
`! git clone https://github.com/atomistic-machine-learning/schnetpack.git -b dev` 
to make sure the dev version is installed. 

I still need to refine some logic behind my initial hypothesis about "PBC is not appropriately guaranteed during either the training or the fitting process of SchNetPack", since I've been worked with only the stable version. Unfortunately, since I used Google Colab as my mean developing platform, I do observed that the dev. version requires Python version >= 3.8, while Colab only supports up to Python 3.7. I decide to move to my HPC resource with multiple CPUs for a demo training, but it seems that work with GPU was the best option. 

@Stefaanhess do you have any advice for using dev version fo SchNetPack on Colab? And I'll of course test with the dev. version with the PaiNN to see if the forces MAEs could be reduced. Any suggestions are appreciate! 

I'll keep in touch during the summer.

Best,
Chenghan Sun
